File: AEDS II/TPs/TP3 (G)/TP03Q1 Lista Sequencial Java/Lista.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Scanner;

public class Lista {
  static Scanner scan= new Scanner(System.in);

  public class Jogador {
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private String anoNasc;
    private String cidadeNasc;
    private String estadoNasc;

    public Jogador() {
      this.id = 0;
      this.nome = "";
      this.altura = 0;
      this.peso = 0;
      this.universidade = "";
      this.anoNasc = "";
      this.cidadeNasc = "";
      this.estadoNasc = "";
    }

    public Jogador(int x) {
      buscarPorId(x, "/tmp/players.csv");
    }

    public int getId() {
      return id;
    }

    public void setId(int id) {
      this.id = id;
    }

    public String getNome() {
      return nome;
    }

    public void setNome(String nome) {
      this.nome = nome;
    }

    public int getAltura() {
      return altura;
    }

    public void setAltura(int altura) {
      this.altura = altura;
    }

    public int getPeso() {
      return peso;
    }

    public void setPeso(int peso) {
      this.peso = peso;
    }

    public String getUniversidade() {
      return universidade;
    }

    public void setUniversidade(String universidade) {
      this.universidade = universidade;
    }

    public String getAnoNasc() {
      return anoNasc;
    }

    public void setAnoNasc(String anoNasc) {
      this.anoNasc = anoNasc;
    }

    public String getCidadeNasc() {
      return cidadeNasc;
    }

    public void setCidadeNasc(String cidadeNasc) {
      this.cidadeNasc = cidadeNasc;
    }

    public String getEstadoNasc() {
      return estadoNasc;
    }

    public void setEstadoNasc(String estadoNasc) {
      this.estadoNasc = estadoNasc;
    }

    protected Jogador clone() throws CloneNotSupportedException {
      Jogador novo = new Jogador();
      novo.id = this.id;
      novo.nome = this.nome;
      novo.altura = this.altura;
      novo.universidade = this.universidade;
      novo.anoNasc = this.anoNasc;
      novo.cidadeNasc = this.cidadeNasc;
      novo.estadoNasc = this.estadoNasc;
      return novo;
    }

    public void ler(String nomeArquivo) throws IOException {
      FileReader file = new FileReader(nomeArquivo);
      BufferedReader buffer = new BufferedReader(file);

      String line;
      boolean firstLine = true;

      while ((line = buffer.readLine()) != null) {
        if (firstLine) {
          firstLine = false;
          continue;
        }

        String[] fields = line.split(",");

        this.id = fields.length > 0 ? Integer.parseInt(fields[0]) : -1;
        this.nome = fields.length > 1 ? fields[1] : "nao informado";
        this.altura = fields.length > 2 ? Integer.parseInt(fields[2]) : -1;
        this.peso = fields.length > 3 ? Integer.parseInt(fields[3]) : -1;
        this.universidade = fields.length > 4 && !fields[4].isEmpty() ? fields[4] : "nao informado";
        this.anoNasc = fields.length > 5 ? fields[5] : "nao informado";
        this.cidadeNasc = fields.length > 6 ? fields[6] : "nao informado";
        this.estadoNasc = fields.length > 7 ? fields[7] : "nao informado";
      }

      buffer.close();
      file.close();
    }

    public void buscarPorId(int playerId, String nomeArquivo) {
      try {
        FileReader file = new FileReader(nomeArquivo);
        BufferedReader buffer = new BufferedReader(file);

        String line;
        boolean firstLine = true;
        // boolean found = false;
        while ((line = buffer.readLine()) != null) {
          if (firstLine) {
            firstLine = false;
            continue;
          }
          String[] fields = line.split(",");

          if (fields.length >= 1) {
            this.id = fields.length > 0 ? Integer.parseInt(fields[0]) : -1;

            if (id == playerId) {
              this.id = fields.length > 0 ? Integer.parseInt(fields[0]) : -1;
              this.nome = fields.length > 1 ? fields[1] : "nao informado";
              this.altura = fields.length > 2 ? Integer.parseInt(fields[2]) : -1;
              this.peso = fields.length > 3 ? Integer.parseInt(fields[3]) : -1;
              this.universidade = fields.length > 4 && !fields[4].isEmpty() ? fields[4] : "nao informado";
              this.anoNasc = fields.length > 5 ? fields[5] : "nao informado";
              this.cidadeNasc = fields.length > 6 ? fields[6] : "nao informado";
              this.estadoNasc = fields.length > 7 ? fields[7] : "nao informado";
              // found = true;
              break;
            }
          }
        }

        buffer.close();
        file.close();

      } catch (Exception jorge) {
        // String toja = "bro";
      }
    }

    public void imprimir() {
      System.out.println(toString());
    }

    public String toString() {
      String resp = "[" + id + " ## " + nome + " ## " + altura + " ## ";
      resp += peso + " ## " + anoNasc + " ## ";
      resp += universidade + " ## " + cidadeNasc + " ## ";
      resp += estadoNasc + "]";
      return resp;
    }

    public void imprimirNome() {
      System.out.println(nome);
    }
  }

  int n, total;
  Jogador[] arr;

  public Lista() {
    n = 0;
    total = 10000;
    arr = new Jogador[total];
  }

  public void inserirInicio(Jogador x){
    this.n++;
    for(int i=n-1;i>=1;i--){
        this.arr[i] = this.arr[i-1];
    }
    this.arr[0] = x;
}

  public void inserirFim(Jogador x) {
    arr[n] = x;
    n++;
  }

  public void removerInicio() {
    System.out.println("(R) " + arr[0].getNome());
    for (int i = 0; i < n - 1; i++) {
      arr[i] = arr[i + 1];
    }
    n--;
  }

  public void removerFim() {
    if (n > 0) {
      System.out.println("(R) " + arr[n - 1].getNome());
      n--;
    }
  }

  public void inserir(Jogador x, int pos){
    for(int i=n-1;i>=pos;i--){
        arr[i+1] = arr[i];
    }
    arr[pos] = x;
    n++;
}

  public void remover(int pos){
    System.out.println("(R) "+arr[pos].getNome());
    for(int i=pos;i<n-1;i++){
        arr[i] = arr[i+1];
    }

    n--;
}

  public void imprimir() {
    for (int i = 0; i < n; i++) {
      arr[i].imprimir();
    }
  }

  public void mostrar() {
    for (int i = 0; i < n; i++) {
        Jogador jogador = arr[i];
        System.out.printf("[%d] ## %s ## %d ## %d ## %s ## %s ## %s ## %s ##%n",
                i, jogador.getNome(), jogador.getAltura(), jogador.getPeso(),
                jogador.getAnoNasc(), jogador.getUniversidade(),
                jogador.getCidadeNasc(), jogador.getEstadoNasc());
    }
}

  public void metodos() {
    String metodo;
    metodo = scan.next();
    if (metodo.equals("II")) {
      int id = scan.nextInt();
      inserirInicio(new Jogador(id));
    }
    if (metodo.equals("IF")) {
      int id = scan.nextInt();
      inserirFim(new Jogador(id));
    }
    if (metodo.equals("I*")) {
      int pos = scan.nextInt();
      int id = scan.nextInt();
      inserir(new Jogador(id), pos);
    }
    if (metodo.equals("RI")) {
      removerInicio();
    }
    if (metodo.equals("RF")) {
      removerFim();
    }
    if (metodo.equals("R*")) {
      int pos = scan.nextInt();
      remover(pos);
    }
  }

  public static void main(String[] args) {
    Lista lista = new Lista();
    String linha;
    linha = scan.nextLine();
    while (!linha.equals("FIM")) {
      int num = Integer.parseInt(linha);
      lista.inserirFim(lista.new Jogador(num));
      linha = scan.nextLine();
    }
    int n = scan.nextInt();
    for (int i = 0; i < n; i++) {
      lista.metodos();
    }
    lista.mostrar();
  }
}

File: AEDS II/TPs/TP3 (G)/TP03Q2 Lista Sequencial C/lista.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Jogador {
  int id;
  char nome[50];
  int altura;
  int peso;
  char universidade[100];
  int anoNasc;
  char cidadeNasc[50];
  char estadoNasc[50];
} Jogador;

void ler(FILE *entrada, Jogador *jogadores) {
  char linha[300];
  fscanf(entrada, "%s", linha);
  char props[8][100];
  int i = 0;
  fgets(linha, 300, entrada);
  while (!feof(entrada)) {
    fgets(linha, 300, entrada);

    int j = 0;
    int a = 0;
    for (int k = 0; k < strlen(linha); k++) {
      if (linha[k] == ',' || linha[k] == '\n') {
        props[j][a] = '\0';
        j++;
        a = 0;
      } else {
        props[j][a] = linha[k];
        a++;
      }
    }

    for (int b = 0; b < 8; b++) {
      if (strcmp(props[b], "") == 0) {
        strcpy(props[b], "nao informado");
      }
    }

    jogadores[i].id = atoi(props[0]);
    strcpy(jogadores[i].nome, props[1]);
    jogadores[i].altura = atoi(props[2]);
    jogadores[i].peso = atoi(props[3]);
    strcpy(jogadores[i].universidade, props[4]);
    jogadores[i].anoNasc = atoi(props[5]);
    strcpy(jogadores[i].cidadeNasc, props[6]);
    strcpy(jogadores[i].estadoNasc, props[7]);

    i++;
  }
  fclose(entrada);
}

void imprimirMensagemRemocao(char *nome) { printf("(R) %s\n", nome); }

typedef struct Lista {
  Jogador *arr;
  int n;
  int total;
} Lista;

Lista *newLista() {
  Lista *lista = (Lista *)malloc(sizeof(Lista));
  lista->arr = (Jogador *)malloc(10000 * sizeof(Jogador));
  lista->n = 0;
  lista->total = 10000;
  return lista;
}

Lista *inserirInicio(Lista *lista, Jogador jogador) {
  for (int i = lista->n; i > 0; i--) {
    lista->arr[i] = lista->arr[i - 1];
  }
  lista->arr[0] = jogador;
  lista->n++;
  return lista;
}

Lista *inserirFim(Lista *lista, Jogador jogador) {
  lista->arr[lista->n] = jogador;
  lista->n++;
  return lista;
}

Lista *removerInicio(Lista *lista) {
  if (lista->n > 0) {
    imprimirMensagemRemocao(lista->arr[0].nome);
    for (int i = 0; i < lista->n - 1; i++) {
      lista->arr[i] = lista->arr[i + 1];
    }
    lista->n--;
  }
  return lista;
}

Lista *removerFim(Lista *lista) {
  if (lista->n > 0) {
    imprimirMensagemRemocao(lista->arr[lista->n - 1].nome);
    lista->n--;
  }
  return lista;
}

Lista *inserir(Lista *lista, Jogador jogador, int pos) {
  for (int i = lista->n; i > pos; i--) {
    lista->arr[i] = lista->arr[i - 1];
  }
  lista->arr[pos] = jogador;
  lista->n++;
  return lista;
}

Lista *remover(Lista *lista, int pos) {
  if (pos >= 0 && pos < lista->n) {
    imprimirMensagemRemocao(lista->arr[pos].nome);
    for (int i = pos; i < lista->n - 1; i++) {
      lista->arr[i] = lista->arr[i + 1];
    }
    lista->n--;
  }
  return lista;
}

void metodos(Lista *lista, Jogador *jogadores) {
  char *metodo = (char *)malloc(3 * sizeof(char));
  scanf("%s", metodo);
  if (strcmp(metodo, "II") == 0) {
    int id;
    scanf("%d", &id);
    inserirInicio(lista, jogadores[id]);
  } else if (strcmp(metodo, "IF") == 0) {
    int id;
    scanf("%d", &id);
    inserirFim(lista, jogadores[id]);
  } else if (strcmp(metodo, "I*") == 0) {
    int id;
    int pos;
    scanf("%d", &pos);
    scanf("%d", &id);
    inserir(lista, jogadores[id], pos);
  } else if (strcmp(metodo, "RI") == 0) {
    removerInicio(lista);
  } else if (strcmp(metodo, "RF") == 0) {
    removerFim(lista);
  } else if (strcmp(metodo, "R*") == 0) {
    int pos;
    scanf("%d", &pos);
    remover(lista, pos);
  }
  free(metodo);
}

void mostrar(Lista *lista) {
  for (int i = 0; i < lista->n; i++) {
    printf("[%d] ## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n", i,
           lista->arr[i].nome, lista->arr[i].altura, lista->arr[i].peso,
           lista->arr[i].anoNasc, lista->arr[i].universidade,
           lista->arr[i].cidadeNasc, lista->arr[i].estadoNasc);
  }
}

int main(void) {
  FILE *entrada = fopen("/tmp/players.csv", "r");
  Jogador *jogadores = (Jogador *)malloc(10000 * sizeof(Jogador));
  Lista *lista = newLista();
  char *linha = (char *)malloc(23 * sizeof(char));
  int n;
  ler(entrada, jogadores);
  scanf("%s", linha);
  while (strcmp(linha, "FIM") != 0) {
    inserirFim(lista, jogadores[atoi(linha)]);
    scanf("%s", linha);
  }
  scanf("%d", &n);
  for (int i = 0; i < n; i++) {
    metodos(lista, jogadores);
  }
  mostrar(lista);
  return 0;
}